File: 207-course-schedule/course-schedule.cs
public class Solution {
    public bool CanFinish(int numCourses, int[][] prerequisites) {
        List<int>[] graph = new List<int>[numCourses];
        for (int i = 0; i < numCourses; i++) {
            graph[i] = new List<int>();
        }

        foreach (var prereq in prerequisites) {
            int course = prereq[0];
            int prerequisite = prereq[1];
            graph[course].Add(prerequisite);
        }

        int[] visited = new int[numCourses];

        for (int i = 0; i < numCourses; i++) {
            if (!DFS(i, graph, visited)) {
                return false; 
            }
        }

        return true; 
    }

    private bool DFS(int course, List<int>[] graph, int[] visited) {
        if (visited[course] == 1) {
            return false;
        }

        if (visited[course] == 2) {
            return true; 
        }

        visited[course] = 1;

        foreach (int prereq in graph[course]) {
            if (!DFS(prereq, graph, visited)) {
                return false; 
            }
        }

        visited[course] = 2;

        return true;
    }
}

File: 207-course-schedule/course-schedule.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function(numCourses, prerequisites) {
    let graph=new Map();
    let inDegree=new Array(numCourses).fill(0);
    for (let  [course, pre] of prerequisites){
        if (!graph.has(pre)) graph.set(pre,[]);
        graph.get(pre).push(course);
        inDegree[course]++;
    }
    let queue=[];
    for (let i=0; i<numCourses; i++){if (inDegree[i]===0)queue.push(i);}
    let takenCourses=0;
    while (queue.length >0){
        let course=queue.shift();
        takenCourses++;
        if(graph.has(course)){
            for (let nextCourse of graph.get(course)){inDegree[nextCourse]--;
            if(inDegree[nextCourse]===0)queue.push(nextCourse);}
        }
    }
    return takenCourses===numCourses;
};